### PR TITLE
Add `CommandExhaustiveLog` class for observing exhaustive optimization

### PR DESCRIPTION
**Pull request by tbirdso**
`CommandExhaustiveLog` extends `itk::Command` to store optimization iteration data in an event-driven manner. Data points are copied to an underlying `itk::Image` with size/spacing/origin generated from `itk::ExhaustiveOptimizerv4` parameters which persists for access once optimization completes.

This PR removes default external module files and adds cxx testing and a Python example notebook.

### CODE REVIEW

**Comment by tbirdso** [comment]
Note `build-windows-python-packages (9)` failures are tracked externally in [scikit-ci-addons](https://github.com/scikit-build/scikit-ci-addons/issues/93).

File: include/itkCommandExhaustiveLog.h [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+#include "itkMacro.h"
+#include "itkCommand.h"
+#include "itkImage.h"
+#include "itkExhaustiveOptimizerv4.h"
+
+namespace itk
+{
+    /**
+     *\class CommandExhaustiveLog
+     *  \brief Command class to collect data from itk::ExhaustiveOptimizerv4 class.
```
Comment (dzenanz) [inline include/itkCommandExhaustiveLog.h]: The introductory paragraph of the example notebook was more useful for understanding the purpose of this class than the documentation here. Perhaps update the docs here?

File: test/itkCommandExhaustiveLogTest.cxx [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+    // Initialize registration
+    registration->SetMetric(metric);
+    registration->SetOptimizer(optimizer);
+    registration->SetFixedImage(fixedImage);
+    registration->SetMovingImage(movingImage);
+    registration->SetInitialTransform(transform);
+    registration->SetNumberOfLevels(1);
+
+    // Create the Command observer and register it with the optimizer.
+    //
```
Comment (dzenanz) [inline test/itkCommandExhaustiveLogTest.cxx]: empty comment line should be removed

File: test/itkCommandExhaustiveLogTest.cxx [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+    ITK_TEST_EXPECT_EQUAL(observer->GetNumberOfSteps(0), 10);
+    ITK_TEST_EXPECT_EQUAL(observer->GetNumberOfSteps(1), 10);
+    ITK_TEST_EXPECT_EQUAL(observer->GetNumberOfSteps(2), 1);
+
+    ITK_TEST_EXPECT_EQUAL(observer->GetStepSize()[0], 0.1);
+    ITK_TEST_EXPECT_EQUAL(observer->GetStepSize()[1], 1.0);
+    ITK_TEST_EXPECT_EQUAL(observer->GetStepSize()[2], 1.0);
+
+    //ITK_TEST_EXPECT_EQUAL(observer->GetDataSize(), 1323);
```
Comment (dzenanz) [inline test/itkCommandExhaustiveLogTest.cxx]: Uncomment or remove?

File: include/itkCommandExhaustiveLog.h (hunk 1 above)
Comment (dzenanz) [reply include/itkCommandExhaustiveLog.h]: Perhaps: `Monitors ExhaustiveOptimizerv4 and provides the search space as an image.`

Comment (tbirdso) [reply include/itkCommandExhaustiveLog.h]: Thanks for the feedback @dzenanz ! I added the brief you suggested and expanded the inline summary.

File: include/itkCommandExhaustiveLog.h [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+  using InternalDataType = TInternalData;
+  /** Image dimension */
+  static constexpr unsigned int Dimension = TImageDimension;
+
+  // Only valid when used with ExhaustiveOptimizerv4 for now
+  // TODO allow different ExhaustiveOptimzerv4 templates
+  // TODO extend for other optimizers
+  using OptimizerType = itk::ExhaustiveOptimizerv4<double>;
+
+  /** Image used as n-dimensional array as a with positional aliasing. */
```
Comment (dzenanz) [inline include/itkCommandExhaustiveLog.h]: as a with -> rephrase

**Review by dzenanz — COMMENTED** [review]
Looks good after a glancing review.

File: include/itkCommandExhaustiveLog.h [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+ * meet their needs.
+ *
+ * Template parameters for class CommandExhaustiveLog:
+ *
+ * - TInternalData = Element type stored at each location in the data image.
+ * - TImageDimension = Dimension of image equal to number of transform parameters
+ *
+ * \ingroup ITKOptimizationMonitor
+ */
+template <typename TInternalData, unsigned int TImageDimension>
```
Comment (thewtex) [inline include/itkCommandExhaustiveLog.h]: To be closer to ITK naming conventions and to be more specific, perhaps `TInternalData` -> `TValue`.

File: include/itkCommandExhaustiveLog.h [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+  void
+  Execute(itk::Object * caller, const itk::EventObject & event) override;
+
+  /** Observe an event fired by calling object. */
+  void
+  Execute(const itk::Object * caller, const itk::EventObject & event) override;
+
+  /** Retrieve a const reference to data point recorded during optimizer iteration. */
+  const TInternalData
+  GetData(const IndexType & index) const;
```
Comment (thewtex) [inline include/itkCommandExhaustiveLog.h]: Instead of `Data`, perhaps the more specific `Value`, to indicate the optimization value for these methods.

**Review by thewtex — CHANGES_REQUESTED** [review]

**Pull request merged**